### ISSUE TITLE
Current Limitations blocking Cypress' full adoption of Webdriver BiDi For Firefox

### ISSUE THREAD

**Issue by AtofStryker** [cypress-io/cypress#30447]
Cypress has spent a few weeks on prepping our firefox automation to cut over from CDP to BiDi. There was some precursor work to do this, such as moving to [geckodriver](https://github.com/cypress-io/cypress/issues/30217) and [webdriver](https://github.com/cypress-io/cypress/issues/30301) in order to have access to a BiDi client on modern automation infrastructure.

Since doing this, we have spiked into cutting over to BiDi, which more details can be seen in the spike issue comment https://github.com/cypress-io/cypress/issues/30351#issuecomment-2432698897.

From what I can see, there are a few blockers for us right now from moving over to BiDi from CDP. This is also being tracked on Firefox's side in https://bugzilla.mozilla.org/show_bug.cgi?id=1604723. 

Priority issues for Cypress in the Bidi Context is cuirrently captured in [Webdriver Bidi Roadmap Planning (Cypress Copy)](https://docs.google.com/spreadsheets/d/1VDnLkrdgJOlta7uiMEQKZFmSrYfpwtBEPsuP3dH5Lmc/edit?gid=0#gid=0). The Cypress team is currently going through issues to prioritize support that we want/need. Original document is [here](https://docs.google.com/spreadsheets/d/1Cg3rifrBZClIitU3aFW_WDv64gY3ge8xPtN-HE1qzrY/edit?gid=0#gid=0
) and will hopefully merge our priorities upstream soon.

Our mapping of CDP/web extension methods to BiDi methods is currently captured in the 
 [Cypress State of Automation](https://docs.google.com/spreadsheets/d/1ED5y68I6RePTTIKLA6SsdP60qlXMUNFtNmNqhdWaeig/edit?gid=0#gid=0) document.

The below checklist is representative from the above link to roadmap planning, reflect P2 issues and under (p1 is more severe than p2)

- [x] P1: Add Support for `resourceType` to `network` events. Issue tracked in https://bugzilla.mozilla.org/show_bug.cgi?id=1904892 and mentioned in https://bugzilla.mozilla.org/show_bug.cgi?id=1604723. This is critical to Cypress to support cookie management, proxy logging, proxy correlation, and the `cy.intercept()` API. UPDATE: December 13th, 2024 This can now be mapped via the `initiatorType` added in https://github.com/w3c/webdriver-bidi/pull/809 which removes the major blocker for us to move to BiDi. This should be available in Firefox 135.
- [ ] P3: File download behavior. The W3C API has `browsingContext.downloadWillBegin`, but this doesn't look to be currently supported as a "no event exist" error is thrown. Cypress also would need something like `Page.downloadComplete` in CDP to fully move this out of our web extension. It looks like this might be fulfilled in https://github.com/w3c/webdriver-bidi/issues/427. We currently work around this in our web extension so it isn't a blocker to base adoption.
- [ ] P3: Cookies onChange behavior, which is tracked to be added to BiDi in https://github.com/w3c/webdriver-bidi/issues/830. We currently work around this in our web extension so it isn't a blocker to base adoption.
- [ ] P3: clearing specific browser data. We accomplish this in CDP via [Storage.clearDataForOrigin](https://vanilla.aslushnikov.com/?Storage.clearDataForOrigin) and [Network.clearBrowserCache](https://vanilla.aslushnikov.com/?Network.clearBrowserCache). This work for BiDi looks to be tracked in https://github.com/w3c/webdriver-bidi/issues/798. We currently work around this in our web extension so it isn't a blocker to base adoption. [storage.deleteCookies](https://www.w3.org/TR/webdriver-bidi/#command-storage-deleteCookies) gets us partially there but not entirely.
- [ ] P2: Be able to load web extensions and access background page. This is currently supported through classic webdriver through firefox specific capabilities. We just want to make sure this is supported in the future if classic webdriver is removed, but isn't a blocker to base adoption.

**Comment by christian-bromann** [comment]
@AtofStryker this is really exciting to see Cypress adopting WebDriver Bidi primitives. If there is anything I can help with from the WebdriverIO project, e.g. any issues you encounter on the `webdriver` or `geckodriver` package, please reach out!

**Comment by AtofStryker** [comment]
> @AtofStryker this is really exciting to see Cypress adopting WebDriver Bidi primitives. If there is anything I can help with from the WebdriverIO project, e.g. any issues you encounter on the `webdriver` or `geckodriver` package, please reach out!

Thank you @christian-bromann. We are mainly blocked on the `resourceType` issue, but once that makes its way into the protocol we will give it another try. I will reach out if we need anything!

**Comment by AtofStryker** [comment]
updated `BIDI_AUTOMATION_CLIENT_NO_RESOURCE_TYPE` (from linked comment in description) on branch `webdriver_ff_no_resource` https://github.com/cypress-io/cypress/compare/release/14.0.0...webdriver_ff_no_resource?expand=1 to get things based off cypress 14 and try `destination`, which was add in firefox nightly 135 as noted in https://bugzilla.mozilla.org/show_bug.cgi?id=1604723#c35. We still need a genuine resource type as destination is only going to tell us where the resource is going, which looks to be right now either a `script`, `style`, `image`, or `iframe`. currently investigating initiator type which has `xmlhttprequest` and `fetch` so it should give us what we need.

Update: initiator type mapping works and is implemented in 072bd68a99 off of `webdriver_ff_no_resource`

**Comment by melibe23** [comment]
With v14 how can we hide network logs like fetch/xhr now?  Thanks!

**Comment by AtofStryker** [comment]
> With v14 how can we hide network logs like fetch/xhr now? Thanks!

@melibe23 you should be able to do this the same way currently. The field is currently deprecated but its still there!

**Comment by adamalston** [comment]
Hi, we use `resourceType` for a few things.

1. Asserting that requests were or weren't sent while performing a set of actions.

```tsx
// cypress/support/index.d.ts
/**
 * Checks requests made during the execution of a set of actions.
 * @param actions A set of actions to perform while intercepting requests.
 * @param requests Expected requests triggered by the actions.
 * @param depth Depth of the comparison for matching requests.
 * - deep: Checks that the intercepted requests and the provided requests
 *   are equal in every way.
 * - shallow: Checks that the intercepted requests and the provided requests
 *   are equal in content and count.
 * - ...
 */
checkRequests(
  actions: () => void,
  requests: { method: string; url: string; }[],
  depth: 'deep' | 'shallow',
): Chainable<Subject>;

// cypress/support/commands.ts
const { baseUrl } = Cypress.config();

Cypress.Commands.add('checkRequests', (actions, requests, depth) => {
  const uuid = crypto.randomUUID();

  cy.intercept({ resourceType: /fetch|xhr/ }).as(uuid);
  actions();

  cy.get<Interception[]>(`@${uuid}.all`, { timeout: 0 }).should(
    depth === 'deep' || depth === 'shallow' ? 'have.length' : 'have.length.gte',
    requests.length,
  );

  switch (depth) {
    case 'deep':
      cy.get<Interception[]>(`@${uuid}.all`).then((interceptions) => {
        cy.wrap(
          interceptions.map(({ request: { method, url } }) => ({
            method,
            url: url.replace(baseUrl ?? '', ''),
          })),
        ).should('deep.eq', requests);
      });
      break;
    // ...
  }
});
```

2. Asserting requests to load fonts.

```tsx
const uuid = crypto.randomUUID();
const expectedRequests = [/* Requests. */];

cy.intercept({
  resourceType:
    /document|fetch|xhr|websocket|stylesheet|script|image|font|cspviolationreport|ping|manifest|other/,
}).as(uuid);

// ...

cy.get<Interception[]>(`@${uuid}.all`).then((interceptions) => {
  const resolvedRequests = expectedRequests.reduce<string[]>(
    (acc, { method, url }) => {
      const resolvedUrl = '/* Processing. */';

      return !resolvedUrl
        ? acc
        : [...acc, JSON.stringify({ method, url: resolvedUrl })];
    },
    [],
  );

  const formattedInterceptions = interceptions.map(
    ({ request: { method, url } }) => JSON.stringify({ method, url }),
  );

  cy.wrap(formattedInterceptions).should('have.members', resolvedRequests);
});
```

Note that I've removed and altered code in the above snippets given the confidential nature of many things I work on.

On https://github.com/cypress-io/cypress-documentation/blob/639cfea75339dd6b617073e99ea2d5661a4010b6/docs/app/references/migration-guide.mdx?plain=1#L98-L103, the two use cases above cover all resource types. Given that, I think they're all important. `fetch` and `xhr` are the most important.

**Comment by jennifer-shehane** [comment]
@adamalston Thanks for the feedback!

**Comment by AtofStryker** [comment]
@adamalston the BiDi changes released yesterday. We are trying to map the resourceTypes the best we can with the new implementation and would appreciate your feedback as to how everything is working!

**Comment by adamalston** [comment]
Do those changes only affect Firefox? I'm asking because we run the vast majority of our tests against Chrome and Edge locally, and Electron in CI. We rarely use Firefox.

**Comment by AtofStryker** [comment]
> Do those changes only affect Firefox? I'm asking because we run the vast majority of our tests against Chrome and Edge locally, and Electron in CI. We rarely use Firefox.

@adamalston the changes only impact Firefox. If for any reason there is an issue, you can force CDP on in firefox with `FORCE_FIREFOX_CDP=1`. More info can be found on https://docs.cypress.io/app/references/launching-browsers#Webdriver-BiDi-and-CDP-Deprecation

**Comment by sergimula** [comment]
Hi there! We use the `resourceType` field to detect XHR calls and add specific headers so our E2E tests can bypass rate limits. If that property goes away and we can’t tell which requests are XHR, we’d have to apply those headers to every request—which definitely isn’t ideal.

**Comment by AtofStryker** [comment]
@sergimula BiDi is able to distinguish the resourceType between `fetch` and `xhr` and that should still be consistent with firefox 135+

**Comment by MikeMcC399** [comment]
@AtofStryker 

Should this issue still be kept open? It seems confusing to have a title "Current Limitations blocking Cypress' adoption of Webdriver BiDi For Firefox" when Cypress already adopted BiDi - see [cypress@14.1.0](https://docs.cypress.io/app/references/changelog#14-1-0) - so they can't be blocking anymore.

If there are non-blocking issues that need to be tracked, then perhaps a new issue would be better?

**Comment by AtofStryker** [comment]
@MikeMcC399 the issue is partially resolved as we were able to do a base adoption of BiDi in Cypress `14.1.0` but not entirely as we still use the [web extension](https://github.com/cypress-io/cypress/blob/v14.5.4/packages/extension/app/v2/background.js) to automate a lot of the Firefox specific tasks. This is largely migrated to BiDi in Cypress 15 with ~90% of the extension code being moved into the BiDi client, but it still doesn't mean we can fully adopt and remove the web extension

Likely a better description would be something like "blocking full adoption of Webdriver BiDi" or something similar that likely hints at removing the web extension?

**Comment by MikeMcC399** [comment]
@AtofStryker 

> Likely a better description would be something like "blocking full adoption of Webdriver BiDi" or something similar that likely hints at removing the web extension?

Thank you very much for explaining! That does all make sense now, so if you were to change from "adoption" to "full adoption" in the title, that would be great!

**Comment by jennifer-shehane** [comment]
Update here from Firefox: https://bugzilla.mozilla.org/show_bug.cgi?id=1604723#c63

>This is now covered and the browsingContext.downloadEnd event is already shipped with Firefox 145.

We'll want to check back in to see if this works and if everything is good with the API.
